### ISSUE TITLE
Nested arrays and dictionaries cause segfault

### ISSUE THREAD

**Issue by Northern-Lights** [plusvic/yara#388]
In the `Writing your own modules` documentation, it is stated that arrays may be nested.
https://yara.readthedocs.org/en/v3.4.0/writingmodules.html

An example given in the documentation is as follows:
```
begin_declarations;

    declare_integer_array("foo");

    begin_struct_array("bar")
        declare_string("baz");
        declare_integer_array("qux");
    end_struct_array("bar");

end_declarations;
```

Examples for accessing these data structures were given as follows:
```
set_integer(<value>, module, "foo[0]");              //Works
set_integer(<value>, module, "foo[%i]", 2);        //Works
set_string(<value>, module, "bar[%i].baz", 5);   //segfaults
set_string(<value>, module, "bar[0].qux[0]");     //set_integer() should be used, but it also segfaults, regardless
set_string(<value>, module, "bar[0].qux[%i]", 0);
set_string(<value>, module, "bar[%i].qux[%i]", 100, 200);
```

It appears that any use of the `bar[]` array with `set_integer()` or `set_string()`, even with a hard-coded index, and even when not accessing the `qux[]` nested array, causes a segfault.

The same segfaulting issue occurs with nested dictionaries, as well.

**Comment by Northern-Lights** [comment]
This should fix it: https://github.com/plusvic/yara/pull/392

**Comment by Northern-Lights** [comment]
There was also an issue where if a `dict` had a nested `dict`, then looking up a value (for an `integer` or something within the struct) that had not been set would cause a segfault because `r2.o` would be UNDEFINED in exec.c.  For example, the following was causing a segfault:
```
import "demo"

rule ddddemo {
condition:
	demo.d1["abc"].i1 == 5 and
	demo.d2["def"].i2 == 5 and
	demo.d2["def"].d2_2["ghi"].i3 == 5 and
	demo.d2["xyz"].d2_2["abc"].i3 > 1
}
```
Where demo.c had these lines:
```
set_integer(5, module_object, "d1[\"abc\"].i1");
set_integer(5, module_object, "d2[\"def\"].i2");
set_integer(5, module_object, "d2[\"def\"].d2_2[\"ghi\"].i3");
```

The yara CLI util would segfault when checking the type of `r2.o` because was UNDEFINED.

My solution: also check for `ensure_defined(r2);` in exec.c in the `OP_LOOKUP_DICT` case.

**Comment by plusvic** [comment]
Fixed in 2ca694b1a2c7b9a3a8d51089a2459ade049516e9
